Sources/TuistKit/Services/CleanService.swift

```diff
@@ -13,20 +13,81 @@ final class CleanService {
         self.cacheDirectoryProviderFactory = cacheDirectoryProviderFactory
     }
 
-    func run(categories: [CacheCategory]) throws {
-        let path: AbsolutePath = FileHandler.shared.currentPath
+    func run(
+        categories: [CleanCategory],
+        path: String?
+    ) throws {
+        let path: AbsolutePath = self.path(path)
         let manifestLoaderFactory = ManifestLoaderFactory()
         let manifestLoader = manifestLoaderFactory.createManifestLoader()
         let configLoader = ConfigLoader(manifestLoader: manifestLoader)
         let config = try configLoader.loadConfig(path: path)
         let cacheDirectoryProvider = try cacheDirectoryProviderFactory.cacheDirectories(config: config)
 
-        for category in categories {
-            let directory = cacheDirectoryProvider.cacheDirectory(for: category)
-            if FileHandler.shared.exists(directory) {
-                try FileHandler.shared.delete(directory)
-                logger.info("Successfully cleaned artifacts at path \(directory.pathString)", metadata: .success)
+        try categories.forEach {
+            switch $0 {
+            case .plugins:
+                try cleanCacheCategory(
+                    .plugins,
+                    cacheDirectoryProvider: cacheDirectoryProvider
+                )
+            case .builds:
+                try cleanCacheCategory(
+                    .builds,
+                    cacheDirectoryProvider: cacheDirectoryProvider
+                )
+            case .tests:
+                try cleanCacheCategory(
+                    .tests,
+                    cacheDirectoryProvider: cacheDirectoryProvider
+                )
+            case .generatedAutomationProjects:
+                try cleanCacheCategory(
+                    .generatedAutomationProjects,
+                    cacheDirectoryProvider: cacheDirectoryProvider
+                )
+            case .projectDescriptionHelpers:
+                try cleanCacheCategory(
+                    .projectDescriptionHelpers,
+                    cacheDirectoryProvider: cacheDirectoryProvider
+                )
+            case .manifests:
+                try cleanCacheCategory(
+                    .manifests,
+                    cacheDirectoryProvider: cacheDirectoryProvider
+                )
+            case .dependencies:
+                try cleanDependencies(at: path)
             }
```

Comment:
Lots of code duplication here. Suggestion: combine `CleanCategory` and `CacheCategory`. Then you can replace this with the following:

```suggestion
            if $0 == .dependencies {
                try cleanDependencies(at: path)
            } else {
                try cleanCacheCategory($0, cacheDirectoryProvider: cacheDirectoryProvider)
            }
```

Reply:
the types won't match if I do it like this, no?

Reply:
I think he is suggestiing to have a single type, which is more or less what I wanted to say here: https://github.com/tuist/tuist/pull/3656#discussion_r750541083 :D

Reply:
If I reused the type, we'd have to pass `path` to [this](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Cache/CacheDirectoriesProvider.swift#L23) call whenever it's used as it's not possible to find `Dependencies` folder without it. That's why I have not reused it in the first place. I can try to refactor it, though, if you feel it makes sense.

Reply:
I think it makes sense, as the Clean is implicitly referring to cleaning the cache (if we consider the local dependencies a cache. If we don't then it should not probably be part of the same command), so I expect the enums to always be aligned

Reply:
> If I reused the type, we'd have to pass `path` to [this](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Cache/CacheDirectoriesProvider.swift#L23) call whenever it's used as it's not possible to find `Dependencies` folder without it. That's why I have not reused it in the first place. I can try to refactor it, though, if you feel it makes sense.

I don't think this is actually an issue. There shouldn't be any code path which can lead to that method being called with a CacheCategory which is `.dependencies`.

Reply:
> if we consider the local dependencies a cache. If we don't then it should not probably be part of the same command)

I tried refactoring it but it makes the `CacheDirectoriesProvider` just awkward to use since you would have to always pass the `path` when the `path` is not available in some contexts. What would make sense is adding a new function  `localCacheDirectory` and `cacheDirectory`. But that's an overkill for this situation because I'd need to add two types `LocalCacheCategory` and `GlobalCacheCategory`. I'd keep it as-is, even though it brings some verbosity to the switch case.

Reply:
How about defining CleanCategory like this?

```swift
/// Category that can be cleaned
enum CleanCategory: CaseIterable, ExpressibleByArgument {
    /// The global cache
    case global(CacheCategory)

    /// The local dependencies cache
    case dependencies

    init?(argument: String) {
        if argument == "dependencies" {
            self = .dependencies
        } else {
            guard let category = CacheCategory(rawValue: argument) else {
                return nil
            }
            self = .global(category)
        }
    }

    static var allCases = CacheCategory.allCases.map { .global($0) } + [Self.dependencies]
}
```

Reply:
That's a good suggestion @ezraberch! Implemented :raised_hands: